### DISCUSSION TITLE
How do you build GrandOrgue from sources, ON WINDOWS.

### DISCUSSION THREAD

**Discussion by wpostma** [GrandOrgue/grandorgue#1126] [Developers]
The problem with cross compiling the WIndows 64 bit GrandOrgue is that it means it's tough or nearly impossible to debug on Windows.

It should be possible to build on Windows via mingw ON WINDOWS but there is no build script provided for that in the build-scripts folder?

**Comment by wpostma** [comment]
Note: I have installed GCC, CMakeNinja, etc, in msys2 on windows and when I build with the $ ./build-scripts/for-win64/build-on-linux.sh
just to see how far it gets, it fails pretty early on at:

> 
> 
> warre@WarrenAlienware MINGW64 /c/audiodev/grandorgue
> $ ./build-scripts/for-win64/build-on-linux.sh
> /c/audiodev/grandorgue/build/build-tools /c/audiodev/grandorgue
> -- Building for: Ninja
> CMake Warning (dev) in CMakeLists.txt:
>   No project() command is present.  The top-level CMakeLists.txt file must
>   contain a literal, direct call to the project() command.  Add a line of
>   code such as
> 
>     project(ProjectName)
> 
>   near the top of the file, but after cmake_minimum_required().
> 
>   CMake is pretending there is a "project(Project)" command on the first
>   line.
> This warning is for project developers.  Use -Wno-dev to suppress it.
> 
> -- The C compiler identification is GNU 11.2.0
> -- The CXX compiler identification is GNU 11.2.0
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - failed
> -- Check for working C compiler: C:/msys64/mingw64/bin/cc.exe
> -- Check for working C compiler: C:/msys64/mingw64/bin/cc.exe - broken
> CMake Error at C:/msys64/mingw64/share/cmake-3.22.3/Modules/CMakeTestCCompiler.cmake:69 (message):
>   The C compiler
> 
>     "C:/msys64/mingw64/bin/cc.exe"
> 
>   is not able to compile a simple test program.
> 
>   It fails with the following output:
> 
>     Change Dir: C:/audiodev/grandorgue/build/build-tools/CMakeFiles/CMakeTmp
> 
>     Run Build Command(s):C:/msys64/usr/bin/ninja.exe cmTC_a05a1 && [1/2] Building C object CMakeFiles/cmTC_a05a1.dir/testCCompiler.c.obj
>     FAILED: CMakeFiles/cmTC_a05a1.dir/testCCompiler.c.obj
>     C:\msys64\mingw64\bin\cc.exe    -o CMakeFiles/cmTC_a05a1.dir/testCCompiler.c.obj -c 'C:/audiodev/grandorgue/build/build-tools/CMakeFiles/CMakeTmp/testCCompiler.c'
>     /bin/sh: line 1: C:msys64mingw64bincc.exe: command not found
>     ninja: build stopped: subcommand failed.
> 
>

**Comment by midimusic** [comment]
In order to build programs on windows you need to forget about Linux and msys2. Instead install mingw64 directly on to windows and use it to compile and install wxwidgets. Then install cmakeGUI. You will also need Portaudio and Rtmidi sources available.

Then try out cmakeGUI on the GO sources. I must admit that I have not done ALL of this myself but I do successfully compile large programs which use wxwidgets on windows. Generally I write my own makefiles as this is much quicker and easier than writing cmake scripts. Though cmake does work well on windows if there is a properly tested and debugged script available.

csw900

**Comment by wpostma** [comment]
So MSYS2 exists for Windows, and as far as I know, is a fork of CygWin GCC, as is MingW64.   I have installed the MSYS2 GCC-windows 64 bit toolchain, but I can try installing Cygwin.

As far as I know "MingW64" is not directly a project name for a Windows installable GCC toolchain. The two choices of WIndows-downloads are MSYS2 and Cygwin.

Gonna try the mingw64 package from cygwin and its support bits ie

<img width="783" alt="image" src="https://user-images.githubusercontent.com/1833470/167224521-842befa6-88b2-4c90-9b36-5dc84f958c56.png">

cygwin items to install:
mingw64 (gcc/g++ for win64)
make
cmake
ninja
 imagemagick
wxWidgets
gettext
docbook-xsl
zip
po4a


jack? (not sure how to install so the cmake pkgconfig rules would pass)

**Comment by wpostma** [comment]
Looks like jack is not found, not sure how to install that on cygwin since cygwin doesn't include a jack devel library for mingw.

$ cmake .
...
-- Checking for module 'jack'
--   Package 'jack', required by 'virtual:world', not found
CMake Error at /usr/share/cmake-3.20.0/Modules/FindPkgConfig.cmake:556 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/share/cmake-3.20.0/Modules/FindPkgConfig.cmake:778 (_pkg_check_modules_internal)
  CMakeLists.txt:133 (pkg_check_modules)


Not sure what to do since pkg-config doesn't seem to be the same kind of thing on cygwin+Windows.  Instead I think some other way must be used to detect the installation of Jack binaries (jackd) or headers or libraries, or compilation options.

$ pkg-config --cflags jack
Package jack was not found in the pkg-config search path.
Perhaps you should add the directory containing `jack.pc'
to the PKG_CONFIG_PATH environment variable
Package 'jack', required by 'virtual:world', not found

**Comment by midimusic** [comment]
You are missing the point - you are still trying to use a Linux Shell running under windows rather than using windows directly. I am not saying that what you are doing will not work but that you are NOT using the basic tools that run DIRECTLY under windows. That is the ones I suggested above and which I know from long experience will work properly.
It is nonsense to say that mingw does not run under windows - I have been using it for many years without any sign of Linux. Also Jack is very seldom used on windows. I always avoid it if possible as I have never found any need for it - even on Linux systems (where it often gets in the way and stops the sound from working properly). GO does not need jack and so you do not have to install it. (You need to give cmake the correct command so that it does not try to use jack - easier with the GUI version I suggested.)

Good Luck csw900

**Comment by wpostma** [comment]
Okay.  Then explain it like I'm five.

What URL do I go to and what thing do I download to install Mingw64 toolchain for windows development?

https://www.mingw-w64.org/downloads/

Because what I figured out for myself is that the mingw downloads page above says to go get msys2 and use that to install MingW.
Then you said, no not msys2.

So then I said, okay then, what?  Cygwin? 

Nope.

Then what?

Is there a third way to get mingw onto my pc (win64)

**Comment by midimusic** [comment]
It is so long ago that I installed mingw on to windows that I have forgotten how I did it. Also things seem to have changed now and the mingw web site seems to have been taken over by pirates. Mingw now appears to be on sourceforge. The link below should get you started. I think you should be looking for an .exe install program.

I started out with mingw32 and this has a built in updater, mingw-get.exe which will list and get all of the components you might  need. I used this for many years to compile 32 bit programs which is no good for GO.  (It can compile 64 bit programs if you specifically tell it what you want but does not do it by default).

Later I wanted to compile 64 bit programs with the least effort so I installed mingw64 which has worked ever since. It still uses parts of mingw32 such as mingw32-make.exe which I regularly use. I now have both installed and the one that gets used depends on which one I select in the Windows PATH directive.

https://sourceforge.net/projects/mingw/files/MinGW/Extension/make/mingw32-make-3.80-3/

Please note that I have not tried this link and as I do not want to upset my computer by running install again so I cannot vouch for it.

I have in the past (now no longer) had both msys2 and cygwin installed on my computer but never found much use for either as I have another  Linux computer if I want to run Linux programs.

csw900
PS. Your link above DOES give an option to download a prebuilt Windows toolchain (as well as lots of others - look about 2/3 of the way down the list.) I have not tried this either so cannot vouch for it.

**Comment by wpostma** [comment]
That sourceforge link is a version of GCC from 2003.

The currently LIVING versions of MingW are MSYS2 and Cygwin+Mingw

What I am trying to say is, MSYS2 does seem to be the way to build with MingW for Win64 in 2022.

What makes you think MSYS2 is NOT the MingW of 2022?

**Comment by midimusic** [comment]
Please yourself what you use but if you compile under msys2 in place of compiling under windows then by default your GO will also "work" only under msys2. As msys2 is not part of Windows and has no Graphic capability your GO probably wont work properly - and you  will be on your own because I don't think anybody else would do it this way. 

I also suspected that Sourceforge might be an archive repository for the sources of mingw. (They probably use github as a repository now.) However the compiled binary version for windows on your link about 2/3 down the page is the one that I would use unless it did not contain a proper windows version of mingw similar to that I regularly use. I assume that this would also conform to your idea of LIVING.

Later:
Having now looked it what this link contains it appears to be sources arranged in a very confused and confusing manner. If this is the best that mingw can now do you may be better off using another compiler. Alternatively do what I understand is the normal way to get a Windows GO - forget about compiling ON windows and cross compile it under Linux. 

csw900

**Comment by wpostma** [comment]
I am not trying to please myself, I'm trying to understand.  I think you're assuming MSYS2 can't be used to create native Win32 apps that use GDI and GL.

Yes, MSYS2 can do that.   As I understand it GrandOrgue uses a library like wx to provide cross platform UI. If the wx libraries and headers are linked into GrandOrge, then the binary GrandOrgue will be the same as a release build for Windows, except that I can debug it now.
  
MingW is just a compiler.

The libraries in MSYS2 do not emulate linux or inject any magic into your executables, as they just give you a gcc exe and its linkers.

Cross compilation is a fine way to get a binary but I can download a binary from the web.

Cross compilation is not a good way to develop and code and debug.

**Comment by midimusic** [comment]
I do not like cross compiling either. But until proved otherwise that is exactly what you are doing with msys2. I will be interested in hearing the outcome of your experiments. Does your msys2 mingw run from a Window command line?

csw900

**Comment by midimusic** [comment]
I have been doing a bit of research into the means of getting a proper windows binary version of mingw-w64.
It is no use searching for this or anything similar as none of the results ever lead to anything useful. The file you need IS on sourceforge but is so well hidden that even google cannot find it without a lot more information.

The file I downloaded "installed" (just copied the files to a directory of my choice) several years ago was:
mingw-w64-bin-i686-mingw_20111220.zip
This is mingw-w64 version 10.3.0 made in 2020 by Brecht Sanders. This file is what you really need but is not currently findable.

However, using my mingw64 - I commanded "gcc --version" from the windows command line and got:
MinGW-W64 x86_64-posix-seh built by Brecht Sanders 10.3.0 (C)2020

If you use google to search for "MinGW-W64 x86_64-posix-seh" without the quotes it will find a similar file on Sourceforge which is compressed with 7z. (The link was so long I could not easily copy it.) This file is actually version 8 about 2 years older but will almost certainly be OK for compiling GO.

PS if you go to https://winlibs.com/ you may find a more up to date version!! (Found by Google: "Brecht Sanders")
csw900

PPS I have just tried mingw64 12.1.0 from winlibs.com on a large program and it compiles OK but will not execute the result because it cannot find an entry point in a .dll . This program compiles and runs OK using my mingw64 V10.3.0 mentioned above. So I would not get too excited about these compilers yet. Best to continue using the older version!!
